freestyle/shared/src/main/scala/freestyle/module.scala

```diff
@@ -16,171 +16,190 @@
-          tyDef1 :: (tyDefs :+ opDef)
-      }
+  private[this] implicit final class ValidateFlatmap[E, A](v: Validated[E, A]) {
+    def flatMap[EE >: E, B](f: A => Validated[EE, B]): Validated[EE, B] =
+      v.andThen(f)
+  }
 
-    // findAlgebras starts from the ClassSymbol of the `@moudle-annotated trait
-    val algebras   = findAlgebras(weakTypeOf[A].typeSymbol.companion.asClass)
-    val coproducts = mkCoproduct(algebras)
-    //ugly hack because, as String,  it does not typecheck, early which we need for types to be in scope
-    val parsed     = coproducts.map( cop => c.parse(cop.toString))
-
-    val expr = if (algebras.length >= 2)
-      q"""new {
-        import _root_.cats.data.Coproduct
-        ..$parsed
-      }"""
+  def impl(annottees: c.Expr[Any]*): Tree =
+    foldAbort(for {
+      cls             <- decodeAnnotationTarget(annottees.toList)
+      effectTuples    <- gatherEffects(cls)
+      classTree       <- makeClassTree(cls)
+      objectTree      <- makeObjectTree(cls, effectTuples)
+    } yield q"$classTree; $objectTree")
+
+  def computeCoproductImpl: Tree =
+    foldAbort(for {
+      tpe             <- compassionateCompanionTypeOf(c.internal.enclosingOwner.owner)
+      fullEffectTypes =  expandEffectType(tpe)
+      fullEffectTrees <- fullEffectTypes.traverse(typeToTypeTree)
+      //coproductTree   <- makeCatsCoproduct(fullEffectTrees) // TODO: make this configurable?
+      coproductTree   <- makeIotaCoproduct(fullEffectTrees)
+    } yield q"""new { ..$coproductTree }""")
+
+  def compassionateCompanionTypeOf(sym: Symbol): ValidatedNel[String, Type] =
+    if (sym.companion.isType)
+      sym.companion.asType.toType.validNel
     else
-      q"""new {
-        ..$parsed
-      }"""
-    c.Expr[Any](expr)
+      s"unable to find companion for $sym when expanding coproduct".invalidNel
+
+  private[this] def decodeAnnotationTarget(annottees: List[c.Expr[Any]]) =
+    annottees.map(_.tree) match {
+      case (cls: ClassDef) :: Nil =>
+        if (cls.mods.hasFlag(Flag.TRAIT | Flag.ABSTRACT)) cls.validNel
+        else
+          "The `@module` annotation can only be applied to a trait or an abstract class.".invalidNel
+
+      case (cls: ClassDef) :: more =>
+        "The trait or class annotated with `@module` must have no companion object.".invalidNel
+      case trees =>
+        "Unexpected trees $trees encountered for `@module` annotation".invalidNel
   }
 
-}
-
-object moduleImpl {
-
-  def impl(c: Context)(annottees: c.Expr[Any]*): c.universe.Tree = {
-    import c.universe._
-    import c.universe.internal.reificationSupport._
+  private[this] def gatherEffects(cls: ClassDef): ValidatedNel[String, List[(Name, Tree)]] =
+    cls.impl.body
+      .collect  { case valDef: ValDef if valDef.rhs.isEmpty => valDef }
+      .traverse { valDef => (valDef.name, valDef.tpt).validNel }
 
-    def fail(msg: String) = c.abort(c.enclosingPosition, msg)
+  private[this] def expandEffectType(tpe: Type): List[Type] = {
```

Comment:
One point in issue #254 is to avoid deep-collecting algebras through transitive dependencies, which removes the need for this recursive procedure. Instead, we would only need to detect which variables in the annotated trait (or class) are `@free` or `@module`, and then just use only the `Op` types of those variables.